File: src/ClaudeToolbar.Core/Mascot/ClawdSprite.cs
namespace ClaudeToolbar.Core.Mascot;

public enum ClawdPose { Rest, Mid, Up, High }

public enum CellKind { Body, Eye }

public readonly record struct SpriteCell(int Col, int Row, CellKind Kind);

/// <summary>Clawd as data: an 18×6 grid of w×2w cells decoded from the Claude Code CLI's block sprite. Row 0 is only used by the raised arm.</summary>
public static class ClawdSprite
{
    public const int Columns = 18;
    public const int Rows = 6;
    public const string BodyColor = "#FFD97757";
    public const string EyeColor = "#FF1B1B1B";
    public const double BadgeCenterCol = 16.5;
    public const double BadgeCenterRow = 0.6;
    public const double BadgeRadiusCells = 2;

    public static ClawdPose PoseFor(double armAngle) =>
        armAngle <= -30 ? ClawdPose.Rest :
        armAngle <= 10 ? ClawdPose.Mid :
        armAngle <= 40 ? ClawdPose.Up : ClawdPose.High;

    public static IReadOnlyList<SpriteCell> Cells(ClawdPose pose)
    {
        var cells = new List<SpriteCell>(56);
        for (var c = 3; c <= 14; c++) cells.Add(new SpriteCell(c, 1, CellKind.Body));
        for (var c = 3; c <= 14; c++) cells.Add(new SpriteCell(c, 2, c is 5 or 12 ? CellKind.Eye : CellKind.Body));
        for (var c = 1; c <= 14; c++) cells.Add(new SpriteCell(c, 3, CellKind.Body));
        for (var c = 3; c <= 14; c++) cells.Add(new SpriteCell(c, 4, CellKind.Body));
        foreach (var c in new[] { 4, 6, 11, 13 }) cells.Add(new SpriteCell(c, 5, CellKind.Body));

        var (inner, outer) = pose switch
        {
            ClawdPose.Rest => (3, 3),
            ClawdPose.Mid => (3, 2),
            ClawdPose.Up => (2, 1),
            _ => (1, 0),
        };
        cells.Add(new SpriteCell(15, inner, CellKind.Body));
        cells.Add(new SpriteCell(16, outer, CellKind.Body));
        return cells;
    }
}

File: mac/Sources/ClaudeToolbarCore/Mascot/ClawdSprite.swift
public enum ClawdPose: Sendable { case rest, mid, up, high }

public enum CellKind: Sendable { case body, eye }

public struct SpriteCell: Hashable, Sendable {
    public let col: Int
    public let row: Int
    public let kind: CellKind

    public init(col: Int, row: Int, kind: CellKind) {
        self.col = col
        self.row = row
        self.kind = kind
    }
}

/// Clawd as data: an 18×6 grid of w×2w cells decoded from the Claude Code CLI's block sprite. Row 0 is only used by the raised arm.
public enum ClawdSprite {
    public static let columns = 18
    public static let rows = 6
    public static let bodyColor = "#FFD97757"
    public static let eyeColor = "#FF1B1B1B"
    public static let badgeCenterCol = 16.5
    public static let badgeCenterRow = 0.6
    public static let badgeRadiusCells = 2.0

    public static func poseFor(armAngle: Double) -> ClawdPose {
        if armAngle <= -30 { return .rest }
        if armAngle <= 10 { return .mid }
        if armAngle <= 40 { return .up }
        return .high
    }

    public static func cells(for pose: ClawdPose) -> [SpriteCell] {
        var cells: [SpriteCell] = []
        cells.reserveCapacity(56)
        for c in 3...14 { cells.append(SpriteCell(col: c, row: 1, kind: .body)) }
        for c in 3...14 { cells.append(SpriteCell(col: c, row: 2, kind: c == 5 || c == 12 ? .eye : .body)) }
        for c in 1...14 { cells.append(SpriteCell(col: c, row: 3, kind: .body)) }
        for c in 3...14 { cells.append(SpriteCell(col: c, row: 4, kind: .body)) }
        for c in [4, 6, 11, 13] { cells.append(SpriteCell(col: c, row: 5, kind: .body)) }

        let (inner, outer): (Int, Int)
        switch pose {
        case .rest: (inner, outer) = (3, 3)
        case .mid: (inner, outer) = (3, 2)
        case .up: (inner, outer) = (2, 1)
        case .high: (inner, outer) = (1, 0)
        }
        cells.append(SpriteCell(col: 15, row: inner, kind: .body))
        cells.append(SpriteCell(col: 16, row: outer, kind: .body))
        return cells
    }
}
